README.md
# cachefun

The goal of `cachefun` is to make caching fun ;-)

It let's you define "cache-aware functions" (`cafun`s or `caf`s): functions that have an internal/built-in cache.

Whether or not you want this cache to be updated upon every execuction of the function or not is completely up to you (see `.refresh_default` argument of `caf_create`):
* When setting `.refresh_default = TRUE` the *caf* would behave just as any regular R function unless you tell it otherwise by setting `.refresh = FALSE` when calling your *caf* 
* When setting `.refresh_default = FALSE`, then it's the other way around: you would always get the value of the internal cache unless you explicitly request the cache to be updated by setting `.refresh = TRUE` when calling your *caf*.

You are also able to define reactive dependencies that directly leverage the reactivity functionality of `{shiny}`. Thus, internal cache values are automatically invalidated if the cache value of one of your function's dependencies changes.

The main use case I had in mind when developping this package was offering effortless ("built-in") **and** flexible (sometimes I don't want the cache result, sometimes I do") iternal caching for functions that take a long time to run (e.g. loading data, tidying data, etc.). 

## Installation

``` r
require("devtools")
devtools::install_github("rappster/cachefun")
```

## Examples

### Introduction 

``` r
library(cachefun)
# Define a regular function that you'd like to make "cache-aware"
fun <- function() Sys.time()

# Turn this function into a cache-aware function
caf <- caf_create(fun = fun)

str(caf)
#> function (..., .fun = function () 
#> Sys.time(), .refresh = TRUE, .reset = FALSE)
# Note that default value for arg '.refresh = TRUE' >> by default, the inner
# function is always executed (internal cache is always updated). This implies
# that the function behaves like any regular R function unless you explicitly
# tell it to return the internal cache in a particular call by setting '.refresh
# = FALSE'.

caf() # Inner function executed, result is cached
#> [1] "2018-02-06 11:05:07 CET"
Sys.sleep(1)
caf() # Inner function executed, result is cached
#> [1] "2018-02-06 11:05:08 CET"
Sys.sleep(1)
caf(.refresh = FALSE) # Inner function NOT executed, internal cache returned
#> [1] "2018-02-06 11:05:08 CET"
Sys.sleep(1)
caf(.refresh = FALSE) # Inner function NOT executed, internal cache returned
#> [1] "2018-02-06 11:05:08 CET"
Sys.sleep(1)
caf() # Inner function executed, result is cached
#> [1] "2018-02-06 11:05:11 CET"
Sys.sleep(1)
caf(.refresh = FALSE) # Inner function NOT executed, internal cache returned
#> [1] "2018-02-06 11:05:11 CET"
```

### Change the default value of args 

``` r
library(cachefun)
fun <- function() Sys.time()
caf <- caf_create(fun = fun, .refresh_default = FALSE)

str(caf)
#> function (..., .fun = function () 
#> Sys.time(), .refresh = FALSE, .reset = FALSE)
# Note that default value for arg '.refresh = FALSE' >> you reversed the
# pre-configured default settings. This implies that the function will always
# return the internal cache value unless you explicitly tell it not to by
# setting `.refresh = TRUE` in a particular call

caf() # Inner function is INITIALLY executed (as cache is still empty),
#> [1] "2018-02-06 11:05:45 CET"
        # result is cached
caf() # Inner function NOT executed, internal cache returned
#> [1] "2018-02-06 11:05:45 CET"
caf(.refresh = TRUE) # Explicit refresh request:
#> [1] "2018-02-06 11:05:45 CET"
                      # Inner function executed, result is cached
caf() # Inner function NOT executed, internal cache returned
#> [1] "2018-02-06 11:05:45 CET"
```

### Inner function with argumeents 

``` r
library(cachefun)
fun <- function(x) Sys.time() + x

caf <- caf_create(fun = fun)

caf(x = 3600) # Inner function executed, result is cached
#> [1] "2018-02-05 10:13:44 CET"
caf(x = 3600 * 5, .refresh = FALSE) # Inner function NOT executed, internal cache returned
#> [1] "2018-02-05 10:13:44 CET"
caf(x = 3600 * 5) # Inner function executed, result is cached
#> [1] "2018-02-05 14:13:44 CET"
```

### Reset internal cache 

``` r
library(cachefun)
fun <- function(x) rnorm(x)

caf <- caf_create(fun = fun)

res <- caf(x = 1000)

caf_reset(caf = caf)
```

### Reactive dependencies

``` r
library(cachefun)

# Define 'caf_1' that 'caf_2' will depend on
fun_1 <- function(x) x
caf_1 <- caf_create(fun = fun_1)

# Define 'caf_2' that depends on 'caf_1'
fun_2 <- function(x, caf_1) {
  caf_1(.refresh = FALSE) + x
}
caf_2 <- caf_create(fun = fun_2, caf_1 = caf_1)
# Note that we state the dependency by relying on the internal cache of 'caf_1'
# ('.refresh = FALSE'). Behind the scenes, 'caf_create' takes care of turning 
# dependencies that are defined as args of the inner function into 
# **reactive** ones (directly leveraging shiny's reactive capabilities). This
# means that 'caf_2' will be re-evaluated whenever the cached return value of
# dependency 'caf_1' is updated. In shiny terms, the cache of 'caf_2' is
# autmatically invalidated when it needs to be and you never run the risk of
# being "out-of-sync" with dependencies

caf_1(x = 10)
#> [1] 10
caf_2(x = 50)
#> [1] 60

caf_1(x = 100)
#> [1] 100
caf_2(x = 50, .refresh = FALSE)
#> [1] 150
# Note that even though we explicitly requested that 'caf_2' should return the
# internally cached value of the previous execution (60), the inner function was
# instead re-evaluated. This is because the cached value of 'caf_1' has changed
# which automatically invalidates all reactive components that depend on it
# (i.e. the cache value of 'caf_2'). This way you can be sure that even though
# you might like to use cached values, you still never run the risk of being
# out-of-sync regarding your function's dependencies

caf_2(x = 200)
#> [1] 300
```


NEWS.md
# cachefun 0.1.0

* Initial version

# cachefun 0.1.1

* Variable default settings
* Reset of internal cache

# cachefun 0.1.2

* Decision on default value of `refresh` in `cafun_create`: it will be `TRUE`

  I think **not** wanting to cache something but instead expecting your cache-aware functions to behave just as "regular" R functions is by far the more typical use case.
  
* Example updated (`inst/examples/ex-cafun_main.R`)

* Update `README.md`
    * Added description and intro text
    * Added example from `inst/examples/ex-cafun_main.R` 

# cachefun 0.1.3

* Implemented observable dependencies (see arg `observes` in `cafun_create`)

# cachefun 0.1.4

* Major changes:
    * Renamed: `cachefun_create` to `caf_create`, `cachefun_reset_cache`  to `caf_reset`.
    * Renamed: distinguish "meta args"" of `caf_create` more explicitly from those of the actual cafuns create by it. Also, probably a good idea to follow the example of `{plyr}` to start meta args with a dot (`.refresh`, `.verbose`).
    * Meta args also aligned in `caf_refresh`.
    * Added `{shiny}` to `Imports`
* Minor changes:
    * Test updated: `test-cafun_main.R` to account for changes.
    * Test added: `test-cafun_main.R:20180205-1`
    * Corrected some minor errors in `inst/examples/ex-cachefun_main.R`.
    * Moved old code for `caf_create` to `R/_scripts/keep_as_reference`.
    * Started to assign explicit IDs to backlog items for easier referencing (see `BACKLOG.md`). The challenge here will be to keep track of the auto-increment state ;-) Think about this a bit more - especially regarding how to leverage/integrate with GitHub issue and labeling system.
    * Updated `README.md`

# cachefun 0.1.5

* Major changes:
    * Improved leveraging of `{shiny}`reactivity (made sure that cache results are automatically invalidated if cache value of dependency changes)
    * Changed argument value in `caf_create`: `observes` is a simple list now
    * Reactive context automatically added behind the scene in `caf_create` if `observes` is supplied (`fun` is wrapped by `reactive`)
    * Added test: 20180206-1 (reactive invalidation) 
* Minor changes:
    * Updated example: inst/examples/ex-cafun_main.R
    * Updated README

# cachefun 0.1.6

* Major changes:
    * Changed argument: `cafun` to `caf` in `caf_reset`
* Minar changes:
    * Changed all occurences of `cafun` to `caf` to make things more concise and consistent

# cachefun 0.1.7

* Major changes:
    * Reactive dependencies (i.e. other *caf*s) can now simply be defined as formal arguments of the inner function. Backend code for handling reactive dependencies has also been consolidated in `caf_create`
    * Three dot argument has been moved up both for `caf_create` and the *caf* created by `caf_create` in order to allow the correct handling of unnamed argument specification when calling the resulting *caf*. This should also make everything play nicely with the tidyverse approach.
* Minor changes:
    * Changed arg `fun` to `.fun` in def of `caf` that is created inside `caf_create` to be more consistent regarding "meta args" (e.g. `.refresh`)
    * Dropped verbose argument(s) (`.verbose` and `.verbose_default`) as I think one can do without them and they just cost compute time.
    * Modified tests in `test-caf_main.R` to reflect changes.
    * Modified examples in `ex-caf_main.R` to reflect changes.


tests/testthat.R
library(testthat)
library(cachefun)

test_check("cachefun")


tests/testthat/test-caf_main.R
library(testthat)

# Basics ------------------------------------------------------------------

context("Cache-aware functions: Basics")

test_that("20180204-1: caf: initial", {
  fun <- function(x) sprintf("hello %s!", x)
  expect_is(caf <- caf_create(fun = fun), "function")

  x_1 <-  "world"
  x_2 <-  "WORLD"

  expectation <- "hello world!"
  expect_identical(caf(x = x_1), expectation)
  expect_identical(caf(x = x_2, .refresh = FALSE), expectation)

  expectation <- "hello WORLD!"
  expect_identical(caf(x = x_2), expectation)
  expect_identical(caf(x = x_1, .refresh = FALSE), expectation)
})

test_that("20180205-1: caf: different default args", {
  fun <- function(x) sprintf("hello %s!", x)
  expect_is(caf <- caf_create(fun = fun, .refresh_default = FALSE), "function")

  x_1 <-  "world"
  x_2 <-  "WORLD"

  expectation <- "hello world!"
  expect_identical(caf(x = x_1), expectation)
  expect_identical(caf(x = x_2), expectation)

  expectation <- "hello WORLD!"
  expect_identical(caf(x = x_2, .refresh = TRUE), expectation)
  expect_identical(caf(x = x_1), expectation)
})

# test_that("20180204-2: caf: verbose", {
#   fun <- function() "hello world!"
#   expect_is(caf <- caf_create(fun = fun), "function")
#
#   expectation <- "hello world!"
#   expect_message(
#     expect_identical(caf(fun = fun, .verbose = TRUE), expectation),
#     "Caching result"
#   )
# })

test_that("20180204-3: caf: reset cache", {
  fun <- function(x) rnorm(x)

  caf <- caf_create(fun = fun)

  res <- caf(x = 3600)

  expect_null(caf_reset(caf = caf))
  # expect_null(caf(.refresh = FALSE))
})

# Reactivity --------------------------------------------------------------

context("Cache-aware functions: reactivity")

test_that("20180204-4: reactivity: basic", {
  fun_1 <- function(x) x
  caf_1 <- caf_create(fun = fun_1)

  fun_2 <- function(x, caf_1) {
    caf_1(.refresh = FALSE) + x
  }
  caf_2 <- caf_create(fun = fun_2, caf_1 = caf_1)

  expect_identical(caf_1(x = 10), 10)
  expect_identical(caf_2(x = 50), 60)
  expect_identical(caf_1(x = 100), 100)
  expect_identical(caf_2(x = 200), 300)
})

test_that("20180206-2: reactivity: invalidation", {
  fun_1 <- function(x) x
  caf_1 <- caf_create(fun = fun_1)

  # fun_2 <- function(x, observes) {
  #   reactive(observes$caf_1(.refresh = FALSE) + x)
  # }
  fun_2 <- function(x, observes) {
    observes$caf_1(.refresh = FALSE) + x
  }
  caf_2 <- caf_create(fun = fun_2, observes = list(caf_1 = caf_1))
  fun_2 <- function(x, caf_1) {
    caf_1(.refresh = FALSE) + x
  }
  caf_2 <- caf_create(fun = fun_2, caf_1 = caf_1)

  caf_1(x = 10)
  expect_identical(caf_2(x = 50, .refresh = FALSE), 60)
  expect_identical(caf_2(x = 50), 60)

  caf_1(x = 100)
  expect_identical(caf_2(x = 50, .refresh = FALSE), 150)
  expect_identical(caf_2(x = 200, .refresh = FALSE), 150)
  expect_identical(caf_2(x = 200), 300)
})


tests/testthat/test-cafun_main.R
library(testthat)

# Basics ------------------------------------------------------------------

context("Cache-aware functions: Basics")

test_that("20180204-1: caf: initial", {
  fun <- function(x) sprintf("hello %s!", x)
  expect_is(caf <- caf_create(fun = fun), "function")

  x_1 <-  "world"
  x_2 <-  "WORLD"

  expectation <- "hello world!"
  expect_identical(caf(x = x_1), expectation)
  expect_identical(caf(x = x_2, .refresh = FALSE), expectation)

  expectation <- "hello WORLD!"
  expect_identical(caf(x = x_2), expectation)
  expect_identical(caf(x = x_1, .refresh = FALSE), expectation)
})

test_that("20180205-1: caf: different default args", {
  fun <- function(x) sprintf("hello %s!", x)
  expect_is(caf <- caf_create(fun = fun, .refresh_default = FALSE), "function")

  x_1 <-  "world"
  x_2 <-  "WORLD"

  expectation <- "hello world!"
  expect_identical(caf(x = x_1), expectation)
  expect_identical(caf(x = x_2), expectation)

  expectation <- "hello WORLD!"
  expect_identical(caf(x = x_2, .refresh = TRUE), expectation)
  expect_identical(caf(x = x_1), expectation)
})

test_that("20180204-2: caf: verbose", {
  fun <- function() "hello world!"
  expect_is(caf <- caf_create(fun = fun), "function")

  expectation <- "hello world!"
  expect_message(
    expect_identical(caf(fun = fun, .verbose = TRUE), expectation),
    "Caching result"
  )
})

test_that("20180204-3: caf: reset cache", {
  # Inner function with argumeents
  fun <- function(x) rnorm(x)

  caf <- caf_create(fun = fun)

  res <- caf(x = 3600)

  caf_reset(caf = caf, .verbose = TRUE)
})

# Reactivity --------------------------------------------------------------

context("Cache-aware functions: reactivity")

test_that("20180204-4: reactivity: basic", {
  fun_1 <- function(x) x
  caf_1 <- caf_create(fun = fun_1)

  fun_2 <- function(x, observes) {
    observes$caf_1(.refresh = FALSE) + x
  }
  caf_2 <- caf_create(fun = fun_2, observes = list(caf_1 = caf_1))

  expect_identical(caf_1(x = 10), 10)
  expect_identical(caf_2(x = 50), 60)
  expect_identical(caf_1(x = 100), 100)
  expect_identical(caf_2(x = 200), 300)
})

test_that("20180206-1: reactivity: invalidation", {

  fun_1 <- function(x) x
  caf_1 <- caf_create(fun = fun_1)

  # fun_2 <- function(x, observes) {
  #   reactive(observes$caf_1(.refresh = FALSE) + x)
  # }
  fun_2 <- function(x, observes) {
    observes$caf_1(.refresh = FALSE) + x
  }

  caf_2 <- caf_create(fun = fun_2, observes = list(caf_1 = caf_1))

  caf_1(x = 10)
  expect_identical(caf_2(x = 50, .refresh = FALSE), 60)
  expect_identical(caf_2(x = 50), 60)

  caf_1(x = 100)
  expect_identical(caf_2(x = 50, .refresh = FALSE), 150)
  expect_identical(caf_2(x = 200, .refresh = FALSE), 150)
  expect_identical(caf_2(x = 200), 300)
})

test_that("20180206-1: reactivity: invalidation", {
  fun_1 <- function(x) x
  caf_1 <- caf_create(fun = fun_1)

  # fun_2 <- function(x, observes) {
  #   reactive(observes$caf_1(.refresh = FALSE) + x)
  # }
  fun_2 <- function(x, observes) {
    observes$caf_1(.refresh = FALSE) + x
  }
  caf_2 <- caf_create(fun = fun_2, observes = list(caf_1 = caf_1))

  caf_1(x = 10)
  expect_identical(caf_2(x = 50, .refresh = FALSE), 60)
  expect_identical(caf_2(x = 50), 60)

  caf_1(x = 100)
  expect_identical(caf_2(x = 50, .refresh = FALSE), 150)
  expect_identical(caf_2(x = 200, .refresh = FALSE), 150)
  expect_identical(caf_2(x = 200), 300)
})


R/auxiliary.R

# Aux ---------------------------------------------------------------------

get_env_to_assign_to <- function(
  use_env = c("temp", "global", "package")
) {
  # nsName <- "temp"
  # (ns <- asNamespace(nsName)) # <environment: namespace:stats>
  # Keep as reference

  # as.environment( "package:dlaker")
  # Keep as reference

  # getPackageName(2)
  # Get name that corresponds to second entry on the search list
  # Keep as reference

  use_env <- match.arg(use_env)

  switch(use_env,
    "temp" = as.environment("._dlaker_temp"),
    "global" = .GlobalEnv,
    # "package" = as.environment(search()[2])

    # TODO-20180131-2:
    # This seems VERY fragile/dangerous and probably only
    # makes sense in context where the datacon generics/methods should be part
    # of a PACKAGE that the developer is building. Not sure what the best
    # option would be in use cases where a developer would work in a PROJECT
    # setting (i.e. only applying packages to solve actual tasks)

    "package" = as.environment(sprintf("package:%s", devtools::as.package(".")))
    # TODO-20180131-2-SOLVED:
    # This seems to be much more robust regarding the package's namespace
    # environment position on the search path
  )
}


R/_scripts/admin.R

# Settings ----------------------------------------------------------------

library(usethis)
repeat_initial <- FALSE

# Misc --------------------------------------------------------------------

use_mit_license("Janko Thyson")
use_roxygen_md()
use_readme_md()
use_news_md()

use_revdep()
# Run checks with `revdepcheck::revdep_check(num_workers = 4)`

# Data --------------------------------------------------------------------

# x <- 1
# y <- 2
# use_data(x, y)
# use_r()

# GitHub credentials ------------------------------------------------------

if (repeat_intial) {
  # cred <- git2r::cred_ssh_key(
  #   normalizePath("~/../.ssh/id_rsa.pub", winslash = "/"),
  #   normalizePath("~/../.ssh/id_rsa", winslash = "/")
  # )
  cred <- git2r::cred_ssh_key(
    "~/.ssh/id_rsa.pub",
    "~/.ssh/id_rsa"
  )
  repo <- git2r::repository(".git")
  git2r::push(repo, credentials = cred)
}

# GitHub ------------------------------------------------------------------

# file.exists("~/../.ssh/id_rsa.pub")
# use_git()
# use_github(credentials = cred)
use_github()

# DevOps ------------------------------------------------------------------

use_travis()
use_coverage()
# covr::package_coverage()


R/internal_cache.R
# Internal cache factory --------------------------------------------------

#' Create cache-aware function
#' @importFrom shiny reactiveValues
#' @importFrom shiny reactive
#' @importFrom shiny isolate
#' @example inst/examples/ex-caf_main.R
#' @export
caf_create <- function(
  fun = NULL,
  ...,
  observes = list(),
  .refresh_default = TRUE
  # .verbose_default = FALSE
) {
  reactv <- shiny::reactiveValues()

  dat_reactive <- shiny::reactive({
    # if(reactv$.verbose) message('Caching result...')
    reactv$data
  })

  caf <- function(
    ...,
    .fun,
    .refresh = TRUE,
    .reset = FALSE
    # .verbose = FALSE,
  ) {
    # Reset -----
    if (.reset) {
      # if (.verbose) message(object.size(dat_reactive()))
      reactv$data <<- NULL
      # if (.verbose) message(object.size(dat_reactive()))
      return(invisible(NULL))
    }

    # Transfer settings -----
    # reactv$.verbose <<- .verbose

    if (is.null(reactv$data) | .refresh) {
      reactv$data <<- reactive(.fun(...))
      # Note:
      # Wrapping with 'reactive' ensures that reactive relationships to
      # dependencies are picked up.
    }

    # Dispatch cache-handling to shiny -----
    if (!shiny::is.reactive(dat_reactive())) {
      dat_reactive()
    } else {
      dat_reactive()()
    }
  }

  # Embedd dependencies as default args of 'fun' -----
  dot_list <- list(...)
  .formals_fun <- formals(fun)
  for (arg in names(dot_list)) {
    .formals_fun[[arg]] <- dot_list[[arg]]
  }
  formals(fun) <- .formals_fun

  # Embedd default values of meta args as args of 'caf'

  # TODO-20180204-1:
  # This seems too inolved >> find better solution

  # TODO-20180205-1:
  # At least encapsulate it in own function

  .formals <- formals(caf)
  if (!is.null(fun)) .formals$.fun <- fun
  .formals$.refresh <- .refresh_default
  # .formals$.verbose <- .verbose_default
  formals(caf) <- .formals

  caf
}

# Reset cache -------------------------------------------------------------

#' Reset internal cache of cache-aware function
#' @example inst/examples/ex-caf_main.R
#' @export
caf_reset <- function(caf, .verbose = FALSE) {
  caf(.reset = TRUE, .verbose = .verbose)
}


inst/examples/ex-cafun_main.R

# Introduction ------------------------------------------------------------

library(cachefun)
# Define a regular function that you'd like to make "cache-aware"
fun <- function() Sys.time()

# Turn this function into a cache-aware function
caf <- caf_create(fun = fun)

str(caf)
# Note that default value for arg '.refresh = TRUE' >> by default, the inner
# function is always executed (internal cache is always updated). This implies
# that the function behaves like any regular R function unless you explicitly
# tell it to return the internal cache in a particular call by setting '.refresh
# = FALSE'.

caf() # Inner function executed, result is cached
Sys.sleep(1)
caf() # Inner function executed, result is cached
Sys.sleep(1)
caf(.refresh = FALSE) # Inner function NOT executed, internal cache returned
Sys.sleep(1)
caf(.refresh = FALSE) # Inner function NOT executed, internal cache returned
Sys.sleep(1)
caf() # Inner function executed, result is cached
Sys.sleep(1)
caf(.refresh = FALSE) # Inner function NOT executed, internal cache returned

# Change the default value of args ----------------------------------------

library(cachefun)
fun <- function() Sys.time()
caf <- caf_create(fun = fun, .refresh_default = FALSE)

str(caf)
# Note that default value for arg '.refresh = FALSE' >> you reversed the
# pre-configured default settings. This implies that the function will always
# return the internal cache value unless you explicitly tell it not to by
# setting `.refresh = TRUE` in a particular call

caf() # Inner function is INITIALLY executed (as cache is still empty),
        # result is cached
caf() # Inner function NOT executed, internal cache returned
caf(.refresh = TRUE) # Explicit refresh request:
                      # Inner function executed, result is cached
caf() # Inner function NOT executed, internal cache returned

# Inner function with arguments ------------------------------------------

library(cachefun)
fun <- function(x) Sys.time() + x

caf <- caf_create(fun = fun)

caf(x = 3600) # Inner function executed, result is cached
caf(x = 3600 * 5, .refresh = FALSE) # Inner function NOT executed, internal cache returned
caf(x = 3600 * 5) # Inner function executed, result is cached

# Reset internal cache ----------------------------------------------------

library(cachefun)
fun <- function(x) rnorm(x)

caf <- caf_create(fun = fun)

res <- caf(x = 1000)

# Resetting the internal cache in verbose mode (messages for prior and new
# object size in cache)
caf_reset(caf = caf, .verbose = TRUE)

# Reactive dependencies ---------------------------------------------------

library(cachefun)

# Define 'caf_1' that 'caf_2' will depend on
fun_1 <- function(x) x
caf_1 <- caf_create(fun = fun_1)

# Define 'caf_2' that depends on 'caf_1'
fun_2 <- function(x, observes) {
  observes$caf_1(.refresh = FALSE) + x
}
caf_2 <- caf_create(fun = fun_2, observes = list(caf_1 = caf_1))
# Note that we state the dependency by relying on the internal cache of 'caf_1'
# ('.refresh = FALSE'). Behind the scenes, 'caf_create' takes care of turning a
# dependency listed in the 'observes' ilist nto an **reactive** one (leveraging
# shiny's capabilities). This means that 'caf_2' will be re-evaluated whenever
# the cached return value of dependency 'caf_1' is updated. In shiny terms, the
# cache of 'caf_2' is autmatically invalidated when it needs to be

caf_1(x = 10)
caf_2(x = 50)

caf_1(x = 100)
caf_2(x = 50, .refresh = FALSE)
# Note that even though we explicitly requested that 'caf_2' should return the
# interanally cached value of the previous execution (60), the inner unction was
# instead evaluated. This is because the cached value of 'caf_1' has changed
# which automatically invalidates all reactive components that depend on it.
# This way you can be sure that even though you might like to use cached values,
# you still never run the risk of being out-of-sync regarding your function's
# dependencies

caf_2(x = 200)


R/_scripts/keep_as_reference/old-caf_create.R

# Keep as reference -------------------------------------------------------

caf_create <- function(
  fun = NULL,
  .refresh = TRUE,
  .verbose = FALSE

) {
  reactv <- shiny::reactiveValues()
  dat_reactive <- shiny::reactive({
    if(reactv$.verbose) message('Caching result...')
    reactv$data
  })

  cafun <- function(
    fun,
    refresh = TRUE,
    reset = FALSE,
    .verbose = FALSE,
    ...
  ) {
    # Reset -----
    if (reset) {
      if (.verbose) message(shiny::isolate(object.size(dat_reactive())))
      # rm(data, envir = reactv)
      reactv$data <<- NULL
      if (.verbose) message(shiny::isolate(object.size(dat_reactive())))
      return(invisible(NULL))
    }

    # Transfer settings -----
    reactv$.verbose <<- .verbose

    if (shiny::isolate(is.null(reactv$data)) | refresh) {
      fun_res <- fun(...)
      # fun_res <- rlang::eval_tidy(fun())
      reactv$data <<- fun_res
    }

    # Relay cache-handling to shiny -----
    shiny::isolate(dat_reactive())
  }

  # Transfer default values -----

  # TODO-20180204-1:
  # This seems too inolved >> find better solution

  .formals <- formals(cafun)
  if (!is.null(fun)) .formals$fun <- fun
  .formals$refresh <- .refresh
  .formals$.verbose <- .verbose
  formals(cafun) <- .formals

  cafun
}


inst/examples/ex-caf_main.R

# Introduction ------------------------------------------------------------

library(cachefun)
# Define a regular function that you'd like to make "cache-aware"
fun <- function() Sys.time()

# Turn this function into a cache-aware function
caf <- caf_create(fun = fun)

str(caf)
# Note that default value for arg '.refresh = TRUE' >> by default, the inner
# function is always executed (internal cache is always updated). This implies
# that the function behaves like any regular R function unless you explicitly
# tell it to return the internal cache in a particular call by setting '.refresh
# = FALSE'.

caf() # Inner function executed, result is cached
Sys.sleep(1)
caf() # Inner function executed, result is cached
Sys.sleep(1)
caf(.refresh = FALSE) # Inner function NOT executed, internal cache returned
Sys.sleep(1)
caf(.refresh = FALSE) # Inner function NOT executed, internal cache returned
Sys.sleep(1)
caf() # Inner function executed, result is cached
Sys.sleep(1)
caf(.refresh = FALSE) # Inner function NOT executed, internal cache returned

# Change the default value of args ----------------------------------------

library(cachefun)
fun <- function() Sys.time()
caf <- caf_create(fun = fun, .refresh_default = FALSE)

str(caf)
# Note that default value for arg '.refresh = FALSE' >> you reversed the
# pre-configured default settings. This implies that the function will always
# return the internal cache value unless you explicitly tell it not to by
# setting `.refresh = TRUE` in a particular call

caf() # Inner function is INITIALLY executed (as cache is still empty),
        # result is cached
caf() # Inner function NOT executed, internal cache returned
caf(.refresh = TRUE) # Explicit refresh request:
                      # Inner function executed, result is cached
caf() # Inner function NOT executed, internal cache returned

# Inner function with arguments ------------------------------------------

library(cachefun)
fun <- function(x) Sys.time() + x

caf <- caf_create(fun = fun)

caf(x = 3600) # Inner function executed, result is cached
caf(x = 3600 * 5, .refresh = FALSE) # Inner function NOT executed, internal cache returned
caf(x = 3600 * 5) # Inner function executed, result is cached

# Reset internal cache ----------------------------------------------------

library(cachefun)
fun <- function(x) rnorm(x)

caf <- caf_create(fun = fun)

res <- caf(x = 1000)

caf_reset(caf = caf)

# Reactive dependencies ---------------------------------------------------

library(cachefun)

# Define 'caf_1' that 'caf_2' will depend on
fun_1 <- function(x) x
caf_1 <- caf_create(fun = fun_1)

# Define 'caf_2' that depends on 'caf_1'
fun_2 <- function(x, caf_1) {
  caf_1(.refresh = FALSE) + x
}
caf_2 <- caf_create(fun = fun_2, caf_1 = caf_1)
# Note that we state the dependency by relying on the internal cache of 'caf_1'
# ('.refresh = FALSE'). Behind the scenes, 'caf_create' takes care of turning a
# dependencies that are defined as args of the inner function into an
# **reactive** ones (directly leveraging shiny's reactive capabilities). This
# means that 'caf_2' will be re-evaluated whenever the cached return value of
# dependency 'caf_1' is updated. In shiny terms, the cache of 'caf_2' is
# autmatically invalidated when it needs to be and you never run the risk of
# being "out-of-sync" with dependencies

caf_1(x = 10)
caf_2(x = 50)

caf_1(x = 100)
caf_2(x = 50, .refresh = FALSE)
# Note that even though we explicitly requested that 'caf_2' should return the
# internally cached value of the previous execution (60), the inner function was
# instead re-evaluated. This is because the cached value of 'caf_1' has changed
# which automatically invalidates all reactive components that depend on it
# (i.e. the cache value of 'caf_2'). This way you can be sure that even though
# you might like to use cached values, you still never run the risk of being
# out-of-sync regarding your function's dependencies

caf_2(x = 200)
